File: pkg/swift/examples/mygit.swift
// mygit.swift — kuu/swift API で構築する git クローンのサンプル
//
// これはモックコードです。kuu Swift ラッパーの API 設計を示すために
// 実際のパース実装なしで「ユーザーがどう書くか」を表現しています。

import Kuu

// MARK: - 共通オプション（全サブコマンドで利用可能）

struct GlobalOptions: ParsableArguments {
    @Count(name: .longAndShort, isGlobal: true, variations: [.reset(prefix: "no")])
    var verbose: Int

    @Flag(name: .longAndShort, isGlobal: true)
    var quiet = false

    @Option(isGlobal: true, choices: ["always", "never", "auto"], implicitValue: "always")
    var color: String = "auto"

    @Flag(name: .customLong("debug-internal"), isGlobal: true, isHidden: true)
    var debugInternal = false

    static var exclusions: [[AnyOption]] {
        [[$verbose, $quiet]]
    }
}

// MARK: - Root Command

struct MyGit: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "mygit",
        abstract: "A sample git-like CLI built with kuu",
        subcommands: [Clone.self, Commit.self, Log.self, Remote.self, Push.self],
        requiresSubcommand: true
    )

    @OptionGroup(isGlobal: true)
    var globals: GlobalOptions
}

// MARK: - clone

struct Clone: ParsableCommand {
    static let configuration = CommandConfiguration(
        abstract: "Clone a repository"
    )

    @OptionGroup(isGlobal: true)
    var globals: GlobalOptions

    // required positional（non-optional = 必須）
    @Argument(help: ArgumentHelp("Repository URL", valueName: "URL"))
    var url: String

    // optional positional（Optional 型 = 任意）
    @Argument(help: "Target directory")
    var directory: String?

    @Option(help: ArgumentHelp("Shallow clone with N commits", valueName: "N"))
    var depth: Int?

    @Option(name: .longAndShort, help: ArgumentHelp("Checkout this branch", valueName: "BRANCH"))
    var branch: String?

    @Flag(help: "Create a bare repository")
    var bare = false

    mutating func run() throws {
        if globals.verbose > 0 {
            print("Verbosity level: \(globals.verbose)")
        }
        print("Cloning \(url)...")
        if let dir = directory {
            print("  into: \(dir)")
        }
        if let d = depth {
            print("  depth: \(d)")
        }
        if let b = branch {
            print("  branch: \(b)")
        }
        if bare {
            print("  bare: true")
        }
    }
}

// MARK: - commit

struct Commit: ParsableCommand {
    static let configuration = CommandConfiguration(
        abstract: "Record changes to the repository"
    )

    @OptionGroup(isGlobal: true)
    var globals: GlobalOptions

    // required option: non-optional + デフォルト値なし
    @Option(name: .longAndShort, help: ArgumentHelp("Commit message", valueName: "MSG"),
            filters: [.trim, .nonEmpty])
    var message: String

    @Flag(name: .longAndShort, help: "Stage all modified files")
    var all = false

    @Flag(help: "Amend the previous commit")
    var amend = false

    // default=true のフラグ。--no-verify で false にする
    @Flag(defaultValue: true, variations: [.false(prefix: "no")],
          help: "Run pre-commit hooks (--no-verify to skip)")
    var verify: Bool

    @Flag(variations: [.unset(prefix: "no")],
          help: "Override author date with committer date")
    var authorDateIsCommitterDate = false

    mutating func run() throws {
        print("Committing: \(message)")
        if all { print("  --all") }
        if amend { print("  --amend") }
        if !verify { print("  --no-verify") }
    }
}

// MARK: - log

struct Log: ParsableCommand {
    static let configuration = CommandConfiguration(
        abstract: "Show commit logs"
    )

    @OptionGroup(isGlobal: true)
    var globals: GlobalOptions

    @Flag(help: "One line per commit")
    var oneline = false

    @Option(name: .longAndShort, help: ArgumentHelp("Limit number of commits", valueName: "N"))
    var maxCount: Int?

    // repeatable option: --author a --author b
    @Repeatable(help: ArgumentHelp("Filter by author (repeatable)", valueName: "PATTERN"))
    var author: [String]

    @Option(choices: ["oneline", "short", "medium", "full", "fuller", "reference", "raw"],
            help: "Pretty-print format")
    var format: String = "medium"

    @Flag(help: "Draw text-based graph of commit history")
    var graph = false

    // -- 以降を path として受け取る
    @Rest(help: "Limit to paths")
    var paths: [String]

    mutating func run() throws {
        print("Showing log")
        if oneline { print("  --oneline") }
        if let n = maxCount { print("  --max-count=\(n)") }
        if !author.isEmpty { print("  --author: \(author)") }
        print("  --format=\(format)")
        if graph { print("  --graph") }
        if !paths.isEmpty { print("  paths: \(paths)") }
    }
}

// MARK: - remote (nested subcommands)

struct Remote: ParsableCommand {
    static let configuration = CommandConfiguration(
        abstract: "Manage set of tracked repositories",
        subcommands: [Add.self, Remove.self],
        requiresSubcommand: true
    )

    @OptionGroup(isGlobal: true)
    var globals: GlobalOptions

    // MARK: remote add

    struct Add: ParsableCommand {
        static let configuration = CommandConfiguration(
            abstract: "Add a remote"
        )

        @OptionGroup(isGlobal: true)
        var globals: GlobalOptions

        @Flag(name: .longAndShort, help: "Fetch after adding")
        var fetch = false

        // serial positional: name, url の順序固定
        @Argument(help: "Remote name")
        var name: String

        @Argument(help: "Remote URL")
        var url: String

        mutating func run() throws {
            print("Adding remote '\(name)' -> \(url)")
            if fetch { print("  --fetch") }
        }
    }

    // MARK: remote remove

    struct Remove: ParsableCommand {
        static let configuration = CommandConfiguration(
            abstract: "Remove a remote"
        )

        @OptionGroup(isGlobal: true)
        var globals: GlobalOptions

        @Argument(help: "Remote name")
        var name: String

        mutating func run() throws {
            print("Removing remote '\(name)'")
        }
    }
}

// MARK: - push

struct Push: ParsableCommand {
    static let configuration = CommandConfiguration(
        abstract: "Update remote refs along with associated objects"
    )

    @OptionGroup(isGlobal: true)
    var globals: GlobalOptions

    @Argument(help: "Remote name")
    var remote: String?

    @Argument(help: "Branch name")
    var branch: String?

    @Flag(name: .longAndShort, help: "Force push")
    var force = false

    @Flag(help: "Safer force push")
    var forceWithLease = false

    @Flag(name: .longAndShort, help: "Set upstream for the branch")
    var setUpstream = false

    // exclusive: --force と --force-with-lease は同時に使えない
    static var exclusions: [[AnyOption]] {
        [[$force, $forceWithLease]]
    }

    mutating func run() throws {
        let remoteName = remote ?? "origin"
        print("Pushing to \(remoteName)")
        if let b = branch { print("  branch: \(b)") }
        if force { print("  --force") }
        if forceWithLease { print("  --force-with-lease") }
        if setUpstream { print("  --set-upstream") }
    }
}

// MARK: - Entry Point

@main
enum Main {
    static func main() {
        MyGit.main()
    }
}

File: pkg/rs/examples/mygit.rs
// mygit — A sample git-like CLI built with kuu-rs to demonstrate all API features.
//
// This is a design mock: it shows the intended API surface, not a compilable program.
// The `kuu` crate and its derive macros do not yet exist; this file serves as the
// specification for their implementation.

use kuu::{Args, KuuError, KuuValue, Subcommand};

// ─── Root CLI ───────────────────────────────────────────────────────────────

#[derive(Args)]
#[kuu(description = "mygit - A sample git-like CLI built with kuu", require_cmd)]
struct Mygit {
    /// Increase verbosity (-v, -vv, -vvv)
    #[kuu(count, short = 'v', global, variation_reset = "no")]
    verbose: i64,

    /// Suppress output
    #[kuu(short = 'q', global, exclusive_group = "verbosity")]
    quiet: bool,

    // verbose is also in the "verbosity" exclusive group
    #[kuu(skip)] // marker: verbose and quiet are exclusive (see below)

    /// When to use colors
    #[kuu(
        default = "auto",
        global,
        choices = ["always", "never", "auto"],
        implicit_value = "always"
    )]
    color: String,

    /// Internal debug flag (hidden from help)
    #[kuu(name = "debug-internal", hidden, global)]
    debug_internal: bool,

    #[kuu(subcommand)]
    command: Command,
}

// exclusive_group は同一グループ名を持つフィールド同士が排他になる。
// count フィールド (verbose) にも exclusive_group を付けたい場合:
//
//   #[kuu(count, short = 'v', global, exclusive_group = "verbosity", variation_reset = "no")]
//   verbose: i64,
//
// ただし count は i64 で flag ではないため、上の例では quiet 側にだけ付けている。
// core 側では p.exclusive([verbose.as_ref(), quiet.as_ref()]) に対応。

// ─── Subcommand Enum ────────────────────────────────────────────────────────

#[derive(Subcommand)]
enum Command {
    /// Clone a repository
    Clone(CloneArgs),

    /// Record changes
    Commit(CommitArgs),

    /// Show commit logs
    Log(LogArgs),

    /// Manage tracked repositories
    Remote(RemoteArgs),

    /// Update remote refs
    Push(PushArgs),
}

// ─── clone ──────────────────────────────────────────────────────────────────

#[derive(Args)]
#[kuu(description = "mygit clone - Clone a repository")]
struct CloneArgs {
    /// Repository URL (required positional)
    #[kuu(positional)]
    url: String,

    /// Target directory (optional positional)
    #[kuu(positional)]
    directory: Option<String>,

    /// Shallow clone with N commits
    #[kuu(value_name = "N")]
    depth: Option<i64>,

    /// Checkout this branch
    #[kuu(short = 'b', value_name = "BRANCH")]
    branch: Option<String>,

    /// Create a bare repository
    bare: bool,
}

// ─── commit ─────────────────────────────────────────────────────────────────

#[derive(Args)]
#[kuu(description = "mygit commit - Record changes")]
struct CommitArgs {
    /// Commit message (required, trimmed, must be non-empty)
    #[kuu(short = 'm', value_name = "MSG", filter = "trim | non_empty")]
    message: String,

    /// Stage all modified files
    #[kuu(short = 'a')]
    all: bool,

    /// Amend the previous commit
    amend: bool,

    /// Run pre-commit hooks (--no-verify to skip)
    #[kuu(default = true, variation_false = "no")]
    verify: bool,
}

// ─── log ────────────────────────────────────────────────────────────────────

#[derive(Args)]
#[kuu(description = "mygit log - Show commit logs")]
struct LogArgs {
    /// One line per commit
    oneline: bool,

    /// Limit number of commits
    #[kuu(short = 'n', value_name = "N", name = "max-count")]
    max_count: Option<i64>,

    /// Filter by author (repeatable: --author A --author B)
    #[kuu(value_name = "PATTERN")]
    author: Vec<String>,

    /// Pretty-print format
    #[kuu(choices = ["oneline", "short", "medium", "full", "fuller", "reference", "raw"])]
    format: Option<LogFormat>,

    /// Limit to paths
    #[kuu(rest)]
    paths: Vec<String>,
}

/// Log output format (auto-generates choices)
#[derive(KuuValue)]
enum LogFormat {
    Oneline,
    Short,
    Medium,
    Full,
    Fuller,
    Reference,
    Raw,
}

// ─── remote (nested subcommands) ────────────────────────────────────────────

#[derive(Args)]
#[kuu(description = "mygit remote - Manage remotes", require_cmd)]
struct RemoteArgs {
    #[kuu(subcommand)]
    command: RemoteCommand,
}

#[derive(Subcommand)]
enum RemoteCommand {
    /// Add a remote
    Add(RemoteAddArgs),

    /// Remove a remote
    Remove(RemoteRemoveArgs),
}

#[derive(Args)]
#[kuu(description = "mygit remote add - Add a new remote", serial)]
struct RemoteAddArgs {
    /// Remote name
    #[kuu(positional)]
    name: String,

    /// Remote URL
    #[kuu(positional)]
    url: String,
}
// serial + never(): #[kuu(serial)] on struct auto-wraps all positionals
// in serial() and appends never() to reject extra arguments.

#[derive(Args)]
#[kuu(description = "mygit remote remove - Remove a remote")]
struct RemoteRemoveArgs {
    /// Remote name
    #[kuu(positional)]
    name: String,
}

// ─── push ───────────────────────────────────────────────────────────────────

#[derive(Args)]
#[kuu(description = "mygit push - Push to remote")]
struct PushArgs {
    /// Force push
    #[kuu(short = 'f', exclusive_group = "force_mode")]
    force: bool,

    /// Safer force push
    #[kuu(name = "force-with-lease", exclusive_group = "force_mode")]
    force_with_lease: bool,

    /// Set upstream for the branch
    #[kuu(short = 'u', name = "set-upstream")]
    set_upstream: bool,
}

// ─── main ───────────────────────────────────────────────────────────────────

fn main() -> Result<(), KuuError> {
    let cli = Mygit::parse()?;

    // Global options are always available
    println!("verbose: {}", cli.verbose);
    println!("quiet: {}", cli.quiet);
    println!("color: {}", cli.color);

    // Type-safe subcommand dispatch — no string matching, no unwrap()
    match cli.command {
        Command::Clone(c) => {
            println!("clone url: {}", c.url); // String, not Option — guaranteed present
            if let Some(dir) = &c.directory {
                println!("clone dir: {}", dir);
            }
            if let Some(depth) = c.depth {
                println!("clone depth: {}", depth);
            }
            if let Some(branch) = &c.branch {
                println!("clone branch: {}", branch);
            }
            println!("clone bare: {}", c.bare);
        }

        Command::Commit(c) => {
            println!("commit message: {}", c.message); // String — filter ensures non-empty
            println!("commit all: {}", c.all);
            println!("commit amend: {}", c.amend);
            println!("commit verify: {}", c.verify); // default=true, --no-verify flips
        }

        Command::Log(l) => {
            println!("log oneline: {}", l.oneline);
            if let Some(n) = l.max_count {
                println!("log max-count: {}", n);
            }
            if !l.author.is_empty() {
                println!("log authors: {:?}", l.author);
            }
            if let Some(fmt) = &l.format {
                println!("log format: {}", fmt);
            }
            if !l.paths.is_empty() {
                println!("log paths: {:?}", l.paths);
            }
        }

        Command::Remote(r) => match r.command {
            RemoteCommand::Add(a) => {
                // serial positionals: name and url are both required
                println!("remote add: {} {}", a.name, a.url);
            }
            RemoteCommand::Remove(r) => {
                println!("remote remove: {}", r.name);
            }
        },

        Command::Push(p) => {
            println!("push force: {}", p.force);
            println!("push force-with-lease: {}", p.force_with_lease);
            println!("push set-upstream: {}", p.set_upstream);
        }
    }

    Ok(())
}
